File: src/test/java/com/example/hexagonal/application/command/CreateOrderHandlerTest.java
package com.example.hexagonal.application.command;

import com.example.hexagonal.adapter.output.FixedIdGenerator;
import com.example.hexagonal.adapter.output.messaging.InMemoryEventPublisher;
import com.example.hexagonal.adapter.output.persistence.InMemoryOrderCommandRepository;
import com.example.hexagonal.adapter.output.persistence.InMemoryOrderQueryRepository;
import com.example.hexagonal.application.dto.CreateOrderInput;
import com.example.hexagonal.application.dto.OrderItemInput;
import com.example.hexagonal.domain.exception.InvalidOrderItemException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.*;

@DisplayName("CreateOrderHandler（Command Handler）")
class CreateOrderHandlerTest {

    private InMemoryOrderCommandRepository commandRepo;
    private InMemoryOrderQueryRepository queryRepo;
    private InMemoryEventPublisher eventPublisher;
    private CreateOrderHandler handler;

    @BeforeEach
    void setUp() {
        commandRepo = new InMemoryOrderCommandRepository();
        queryRepo = new InMemoryOrderQueryRepository();
        eventPublisher = new InMemoryEventPublisher(commandRepo, queryRepo);
        handler = new CreateOrderHandler(commandRepo, eventPublisher, new FixedIdGenerator("order-1"));
    }

    @Test
    void 應該成功建立訂單() {
        var input = new CreateOrderInput("c1", List.of(
                new OrderItemInput("p1", "鍵盤", 2500, 1)
        ));

        var output = handler.execute(input);

        assertThat(output.orderId()).isEqualTo("order-1");
        assertThat(output.totalAmount()).isEqualTo(2500);
        assertThat(output.status()).isEqualTo("DRAFT");
        assertThat(output.itemCount()).isEqualTo(1);
    }

    @Test
    void 應該將訂單儲存到命令儲存庫() {
        var input = new CreateOrderInput("c1", List.of(
                new OrderItemInput("p1", "鍵盤", 2500, 1)
        ));

        handler.execute(input);

        assertThat(commandRepo.size()).isEqualTo(1);
        assertThat(commandRepo.findById("order-1")).isPresent();
    }

    @Test
    void 應該發布領域事件() {
        var input = new CreateOrderInput("c1", List.of(
                new OrderItemInput("p1", "鍵盤", 2500, 1)
        ));

        handler.execute(input);

        // OrderCreated + OrderItemAdded
        assertThat(eventPublisher.getPublishedEvents()).hasSizeGreaterThanOrEqualTo(2);
    }

    @Test
    void 應該同步讀取模型_CQRS查詢端() {
        var input = new CreateOrderInput("c1", List.of(
                new OrderItemInput("p1", "鍵盤", 2500, 1),
                new OrderItemInput("p2", "滑鼠", 800, 2)
        ));

        handler.execute(input);

        var readModel = queryRepo.findById("order-1");
        assertThat(readModel).isPresent();
        assertThat(readModel.get().totalAmount()).isEqualTo(4100);
        assertThat(readModel.get().items()).hasSize(2);
    }

    @Test
    void 無效的訂單項目應該拋出錯誤() {
        var input = new CreateOrderInput("c1", List.of(
                new OrderItemInput("", "鍵盤", 2500, 1)
        ));

        assertThatThrownBy(() -> handler.execute(input))
                .isInstanceOf(InvalidOrderItemException.class);
    }
}

File: tests/application/CreateOrderHandler.test.ts
// ============================================================
// Command Handler 測試 - 建立訂單
// ============================================================
// 應用層測試展示了六角形架構的可測試性：
// 我們用簡單的記憶體實作替換外部依賴，
// 完全不需要資料庫或訊息佇列。
//
// 這就是 DIP（依賴反轉原則）帶來的好處。
// ============================================================

import { CreateOrderHandler } from '../../src/application/commands/CreateOrderHandler';
import { InMemoryOrderCommandRepository } from '../../src/adapters/output/persistence/InMemoryOrderCommandRepository';
import { InMemoryOrderQueryRepository } from '../../src/adapters/output/persistence/InMemoryOrderQueryRepository';
import { InMemoryEventPublisher } from '../../src/adapters/output/messaging/InMemoryEventPublisher';
import { FixedIdGenerator } from '../../src/adapters/output/IdGeneratorAdapter';

describe('CreateOrderHandler', () => {
  let handler: CreateOrderHandler;
  let commandRepo: InMemoryOrderCommandRepository;
  let queryRepo: InMemoryOrderQueryRepository;
  let eventPublisher: InMemoryEventPublisher;

  beforeEach(() => {
    commandRepo = new InMemoryOrderCommandRepository();
    queryRepo = new InMemoryOrderQueryRepository();
    eventPublisher = new InMemoryEventPublisher(commandRepo, queryRepo);
    const idGenerator = new FixedIdGenerator('test-order-001');

    handler = new CreateOrderHandler(commandRepo, eventPublisher, idGenerator);
  });

  it('應該成功建立訂單', async () => {
    const result = await handler.execute({
      customerId: 'customer-1',
      items: [
        { productId: 'p1', productName: '鍵盤', unitPrice: 2500, quantity: 1 },
        { productId: 'p2', productName: '滑鼠', unitPrice: 800, quantity: 2 },
      ],
    });

    expect(result.orderId).toBe('test-order-001');
    expect(result.totalAmount).toBe(4100);
    expect(result.status).toBe('DRAFT');
    expect(result.itemCount).toBe(2);
  });

  it('應該將訂單儲存到命令儲存庫', async () => {
    await handler.execute({
      customerId: 'customer-1',
      items: [{ productId: 'p1', productName: '鍵盤', unitPrice: 2500, quantity: 1 }],
    });

    const saved = await commandRepo.findById('test-order-001');
    expect(saved).not.toBeNull();
    expect(saved!.customerId).toBe('customer-1');
  });

  it('應該發布領域事件', async () => {
    await handler.execute({
      customerId: 'customer-1',
      items: [{ productId: 'p1', productName: '鍵盤', unitPrice: 2500, quantity: 1 }],
    });

    const events = eventPublisher.getPublishedEvents();
    expect(events.length).toBeGreaterThan(0);

    const eventTypes = events.map((e) => e.eventType);
    expect(eventTypes).toContain('OrderCreated');
    expect(eventTypes).toContain('OrderItemAdded');
  });

  it('應該同步讀取模型（CQRS 查詢端）', async () => {
    await handler.execute({
      customerId: 'customer-1',
      items: [{ productId: 'p1', productName: '鍵盤', unitPrice: 2500, quantity: 1 }],
    });

    const readModel = await queryRepo.findById('test-order-001');
    expect(readModel).not.toBeNull();
    expect(readModel!.totalAmount).toBe(2500);
    expect(readModel!.items).toHaveLength(1);
  });

  it('無效的訂單項目應該拋出錯誤', async () => {
    await expect(
      handler.execute({
        customerId: 'customer-1',
        items: [{ productId: '', productName: '鍵盤', unitPrice: 2500, quantity: 1 }],
      }),
    ).rejects.toThrow();
  });
});
